src/java.desktop/share/classes/javax/swing/text/html/CSS.java

```diff
@@ -2200,6 +2213,17 @@ Object toStyleConstants(StyleConstants key, View v) {
             return Integer.valueOf(getValue(null, null));
         }
 
+        @Override
+        public int hashCode() {
+            return Float.hashCode(value);
+        }
+
+        @Override
+        public boolean equals(Object val) {
+                return val instanceof CSS.FontSize size
+                       && Objects.equals(size.svalue, svalue);
+        }
```

Comment:
This is an anomaly. In all the other cases below you are using the local fields not the original parsed string.

Reply:
This is because font-size attribute can have units or percentage in it, it seems like as is being tested in the regression test, 42px, 42pt so "svalue" is being used which will have 42px/42pt so we can compare the string value of 42px or 42pt otherwise we need to parse to get the integer 42 and then the units px or pt and then compared separately

Reply:
We should not parse the string, it's already parsed.

What if I put a space between value and unit? These two strings should produce equal attribute sets:
```java
{"font-size: 42px", "font-size: 42 px"},
```

The easiest way is to implement `LengthUnit.equals` which you still need for comparing `CSS.LengthValue` anyway because that class has exactly the same problem. You also compare the raw string there:

https://github.com/openjdk/jdk/blob/612d5eba4c83ab53c797b03a244879261a5faa2a/src/java.desktop/share/classes/javax/swing/text/html/CSS.java#L2678-L2681

It doesn't look right to me.

This is basically [a continuation of the same discussion](https://github.com/openjdk/jdk/pull/13405#discussion_r1202100159).

Reply:
I am not sure I understand...Where the string is parsed...Earlier in my previous iteration I was using Scanner util class to parse the string to extract integer 42 and px from 42px and comparing separately which I thought you were referring to as we should not parse in `equals` so I removed and I optimized to compare the string value `svalue`

What are you suggesting then? I already told using your suggested code fails the regression test..

Reply:
> I already told using your suggested code fails the regression test..

Which regression test do you refer to? The one for `AttributeSet.equals` or another?

I can't see how implementing `equals` could affect other tests… unless some tests already use `AttributeSet.equals`.

If the test being developed fails, it means the implementation in `CSS` is wrong.

I haven't been able to thinker with the code myself.

> What are you suggesting then?

Neither parsing the string nor using the raw string seems right. The object already contains the parsed data as a set of fields: `value`, `index`, `lu`.
https://github.com/openjdk/jdk/blob/927a9ed68371597eba1609f97ac03dd1de812e26/src/java.desktop/share/classes/javax/swing/text/html/CSS.java#L2203-L2205

I think the fields should be used in the implementation of the `equals` method, and, as [I already noted](https://github.com/openjdk/jdk/pull/13405#discussion_r1200351364), it should look something like this:
```java
        @Override
        public boolean equals(Object val) {
            return val instanceof CSS.FontSize size
                   && value == size.value
                   && index == size.index
                   && Objects.equals(lu, size.lu);
        }
```

Using `Objects.equals` handles `null` values in `lu` fields gracefully. At the same time, this requires that `LengthUnit` implements `equals`. I suggested an implementation without `LengthUnit.equals`, it was probably incorrect because I didn't test it; however, implementing `LengthUnit.equals` is a better solution.

```java
        @Override
        public boolean equals(Object obj) {
            return obj instanceof LengthUnit ou
                   && type == ou.type
                   && value == ou.value
                   && Objects.equals(units, ou.units);
        }
```

(Okay, I thought `CSS.LengthValue` also uses `LengthUnit` class but it doesn't, or rather it doesn't have it as a field, units are stored as a `String` value there.)